btree/btree.c

```diff
@@ -2939,145 +2953,128 @@ static void ff_capture(struct slot *slot, struct m0_be_tx *tx)
+static void ff_node_update_credit(const struct nd *node,
+				   struct m0_be_tx_credit *cred)
 {
-	accum->tc_cb_nr += BTREE_CALLBACK_CREDIT;
+	int         shift     = ff_shift(node);
+	m0_bcount_t node_size = 1ULL << shift;
+
+	m0_be_tx_credit_add(cred, &M0_BE_TX_CREDIT(1, node_size));
 }
 
 /**
- * This function will calculate credits required to allocate node and it will
- * add those credits to @accum.
+ * This function will calculate credits required to free the node and it will
+ *  add those credits to @cred.
  */
-static void btree_node_alloc_credit(const struct m0_btree  *tree,
-				    struct m0_be_tx_credit *accum)
+static void ff_node_free_credit(const struct nd *node,
+				struct m0_be_tx_credit *cred)
 {
-	m0_bcount_t             node_size;
-	int                     shift;
+	struct ff_head *h           = ff_data(node);
+	int             shift       = h->ff_shift;
+	int             node_size   = 1ULL << shift;
+	int             header_size = sizeof(*h);
 
-	shift     = node_shift(tree->t_desc->t_root);
-	node_size =  1ULL << shift;
+	m0_be_allocator_credit(NULL, M0_BAO_FREE_ALIGNED,
+			       node_size, shift, cred);
 
-	m0_be_allocator_credit(NULL, M0_BAO_ALLOC_ALIGNED,
-			       node_size, shift, accum);
+	m0_be_tx_credit_add(cred, &M0_BE_TX_CREDIT(1, header_size));
 }
 
 /**
- * This function will calculate credits required to update node and it will add
+ * This function will calculate credits required to split node and it will add
  * those credits to @accum.
  */
-static void btree_node_update_credit(const struct m0_btree  *tree,
-				     struct m0_be_tx_credit *accum)
+static void ff_node_split_credit(const struct nd *node,
+				 struct m0_be_tx_credit *cred)
 {
- 	m0_bcount_t             node_size;
-	int                     shift;
+	struct m0_be_tx_credit cred_update_node = {};
 
-	shift     = node_shift(tree->t_desc->t_root);
-	node_size =  1ULL << shift;
+	ff_node_alloc_credit(node, cred);
 
-	m0_be_tx_credit_add(accum, &M0_BE_TX_CREDIT(1, node_size));
-}
+	/* credits to update two nodes : existing and newly allocated. */
+	ff_node_update_credit(node, &cred_update_node);
+	m0_be_tx_credit_mul(&cred_update_node, 2);
 
+	m0_be_tx_credit_add(cred, &cred_update_node);
+}
 
-/**
- * This function will calculate credits required for the delete KV operation and
- * add those credits to @accum.
- */
-static void btree_del_credit(const struct m0_btree  *tree, m0_bcount_t ksize,
-			     m0_bcount_t vsize, struct m0_be_tx_credit *accum)
+static void ff_put_credit(const struct nd *node, m0_bcount_t ksize,
+			  m0_bcount_t vsize, struct m0_be_tx_credit *cred)
 {
-	struct m0_be_tx_credit cred = {};
-
-	/* Credits for freeing the node. */
-	node_free_credits(tree->t_desc->t_root, ksize, vsize, &cred);
-	m0_be_tx_credit_mac(accum, &cred, MAX_TREE_HEIGHT);
+	struct m0_be_tx_credit put_cred = {};
+	/* Credits for split operation */
+	ff_node_split_credit(node, &put_cred);
+	m0_be_tx_credit_mac(cred, &put_cred, MAX_TREE_HEIGHT);
```

Comment:
MAX_TREE_HEIGHE is not known to ff_* routines. The ff_put_credit() only gives the credit count for this node. The rest of the addition/multiplication needs to be done by the caller routine - m0_put_credit() which knows about the tree height and also any tree related changes happening as a part of put_tick().